File: src/semester_SQL_creator/CourseInfoFinder.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.net.URI;
import java.net.http.HttpClient;
import java.net.http.HttpRequest;
import java.net.http.HttpResponse;
import java.util.HashMap;
import java.util.Vector;
import org.json.JSONArray;
import org.json.JSONObject;

// Compile: javac -cp "lib/*" CourseInfoFinder.java
// Run: java -cp ".;lib/*" CourseInfoFinder

public class CourseInfoFinder {

    static String[] departments = new String[]{"AHIS", "ALI", "AMST", "ANTH", "ARAB", "ASTR", "BISC", "CHEM", "CLAS", "COLT", "CORE", "CSLC",
               "EALC", "EASC", "ECON", "ENGL", "ENST", "FREN", "GEOG", "GEOL", "GERM", "SWMS", "GR", "HEBR",
               "HIST", "HBIO", "INDS", "IR", "IRAN", "ITAL", "JS", "LAT", "LING", "MATH", "MDA", "MDES", "MPW",
               "NEUR", "NSCI", "OS", "PHED", "PHIL", "PHYS", "POIR", "PORT", "POSC", "PSYC", "REL", "RNR",
               "SLL", "SOCI", "SPAN", "SSCI", "SSEM", "USC", "VISS", "WRIT", "ACCT", "ARCH", "ACAD", "ACCT",
               "BAEP", "BUAD", "BUCO", "DSO", "FBE", "GSBA", "MKT", "MOR", "HRM", "CMPP", "CNTV", "CTAN",
               "CTCS", "CTIN", "CTPR", "CTWR", "IML", "ASCJ", "CMGT", "COMM", "DSM", "JOUR", "PR", "PUBD",
               "DANC", "DENT", "CBY", "DHIS", "THTR", "EDCO", "EDHP", "EDUC", "AME", "ASTE", "BME", "CHE", "CE",
               "CSCI", "EE", "ENE", "ENGR", "ISE", "INF", "ITP", "MASC", "PTE", "SAE", "ART", "CRIT", "DES",
               "FA", "FACE", "FACS", "FADN", "FADW", "FAIN", "FAPH", "FAPT", "FAPR", "FASC", "WCT", "GCT",
               "SCIN", "SCIS", "ARLT", "SI", "ARTS", "HINQ", "SANA", "LIFE", "PSC", "QREA", "GPG", "GPH",
               "GESM", "GERO", "LAW", "ACMD", "ANST", "BIOC", "CBG", "DSR", "HP", "INTD", "MEDB", "MEDS",
               "MICB", "MPHY", "MSS", "NIIN", "PATH", "PHBI", "PM", "PCPA", "SCRM", "ARTL", "MTEC", "MSCR",
               "MTAL", "MUCM", "MUCO", "MUCD", "MUEN", "MUHL", "MUIN", "MUJZ", "MPEM", "MPGU", "MPKS", "MPPM",
               "MPST", "MPVA", "MPWP", "MUSC", "SCOR", "OT", "HCDA", "PHRD", "PMEP", "RXRS", "BKN", "PT",
               "AEST", "HMGT", "MS", "NAUT", "NSC", "PPD", "PPDE", "PLUS", "RED"};

    // Note: 20243 = Fall 2024
    static String semester = "20243";

    private static String getDepartmentJsonString(String department_id, String semester_id) throws Exception{
        try {
            String url_text = "https://web-app.usc.edu/web/soc/api/classes/" + department_id + "/" + semester_id;
            
            // Create a HttpClient
            HttpClient client = HttpClient.newHttpClient();
            
            // Create a HttpRequest
            HttpRequest request = HttpRequest.newBuilder().uri(URI.create(url_text)).build();
            
            // Send the request and get the response
            HttpResponse<String> response = client.send(request, HttpResponse.BodyHandlers.ofString());
            return response.body();
            
        } catch (IOException | InterruptedException e) {
            throw e;
        }

    }
    public static Vector<Course> createClassMapping(String json_response){
        Vector<Course> courses = new Vector<Course>();
        HashMap<String, Course> nameToCourse = new HashMap<String, Course>();

        // Parse the JSON string into a JSONObject
        JSONObject jsonObject = new JSONObject(json_response);
        JSONArray all_courses = (jsonObject.getJSONObject("OfferedCourses")).getJSONArray("course");
        for(int i = 0; i < all_courses.length(); i++){
           
            JSONObject course_data = ((JSONObject) all_courses.get(i)).getJSONObject("CourseData");
            
            String name = course_data.getString("prefix") + "-" + course_data.getString("number");

            String title;
            try{
                title = course_data.getString("title");
            }
            catch(Exception e){
                title = "N/A";
            }

            String description = "N/A";
            try{
                description = course_data.getString("description");
            }
            catch(Exception e){
                description = "N/A";
            }

            String units;
            try{
                units = course_data.getString("units");
                units = units.split(",")[0];
            }
            catch(Exception e){
                units = "N/A";
            }

            // Add course to courses if it doesn't exist
            if(!(nameToCourse.containsKey(name))){
                Vector<Section> sections = new Vector<Section>();
                Course c = new Course(name, title, description, units, sections);
                nameToCourse.put(name, c); // update hashmap
                courses.addElement(c); // add to list of courses
            }

            JSONArray all_sections;
            try{
                all_sections = course_data.getJSONArray("SectionData");
                
            }
            catch(Exception e){
                // only one section
                
                JSONObject section = course_data.getJSONObject("SectionData");


                all_sections = new JSONArray();

            }
            

            for(int j = 0; j < all_sections.length(); j++){
                JSONObject section_data = ((JSONObject) all_sections.get(j));

                String id_and_d_class_code;
                try{
                    id_and_d_class_code = section_data.getString("id") + section_data.getString("dclass_code");
                }
                catch(Exception e){
                    id_and_d_class_code = "Unknown";
                }

                String type;
                try{
                    type = section_data.getString("type");
                }
                catch(Exception e){
                    type = "Unknown";
                }

                String day;
                try{
                    day = section_data.getString("day");
                }
                catch(Exception e){
                    day = "Unknown";
                }

                // Create the time string
                String time;
                try{
                    time = section_data.getString("start_time") + "-" + section_data.getString("end_time");
                }
                catch(Exception e){
                    time = "Unknown";
                }

                String loc;
                try{
                    loc = section_data.getString("location");
                }
                catch(Exception e){
                    loc = "Unknown";
                }

                String prof_name;
                try{
                    JSONObject prof = section_data.getJSONObject("instructor");
                    prof_name = prof.getString("last_name") + "," + prof.getString("first_name");
                }
                catch(Exception e){
                    prof_name = "Unknown";
                }
                
                

                System.out.println(id_and_d_class_code);
                System.out.println(type);
                System.out.println(day);
                System.out.println(time);
                System.out.println(loc);
                System.out.println(prof_name);
                Section s = new Section(id_and_d_class_code, type, day, time, loc, prof_name);
                (nameToCourse.get(name)).sections.addElement(s);

            }

            //System.out.println(all_sections);

            /*
            System.out.println(name);
            System.out.println(title);
            System.out.println(description);
            System.out.println(units);
            */

            
        }
        
        System.out.println();
        return courses;
    }

    public static void printCourses(Vector<Course> courses){
        for(int i = 0; i < courses.size(); i++){
            Course c = courses.get(i);
            System.out.println("Name: " + c.name);
            System.out.println("Title: " + c.title);
            System.out.println("Description: " + c.description);
            System.out.println("--------------------------------------------------");

            for(int j = 0; j < c.sections.size(); j++){
                Section s = c.sections.get(j);
                System.out.println("ID: " + s.id_and_d_class_code);
                System.out.println("Type: " + s.type);
                System.out.println("Day: " + s.day);
                System.out.println("Time: " + s.time);
                System.out.println("Location: " + s.loc);
                System.out.println("Professor: " + s.prof_name);
                System.out.println();
            }

        }
           
        
    }
    
    public static void main(String[] args){
        for(int i = 0; i < departments.length; i++){
            String department = departments[i];
            try{
                String json_response = getDepartmentJsonString(department, semester);
                System.out.println(json_response);
                Vector<Course> courses = createClassMapping(json_response);
                printCourses(courses);
            }
            catch(Exception e){
                System.out.println("Error while fetching data for " + department + ": " + e.getMessage());
            }
        }
        
    }
}

class Section{
    String id_and_d_class_code;
    String type;
    String day;
    String time;
    String loc; 
    String prof_name;
    Section(String id_and_d_class_code, String type, String day, String time, String loc, String prof_name){
        this.id_and_d_class_code = id_and_d_class_code;
        this.type = type; 
        this.day = day;
        this.time = time;
        this.loc = loc;
        this.prof_name = prof_name;
    }
}

class Course{
    String name;
    String title;
    String description; 
    String units;
    Vector<Section> sections;

    Course(String name, String title, String description, String units, Vector<Section> sections){
        this.name = name;
        this.title = title;
        this.description = description; 
        this.sections = sections;
    }
}

File: pull current semester course info into base/depricated_python/courseInfoFinder.py
import json
import requests
from dataclasses import dataclass
from typing import List

departments = ["AHIS", "ALI", "AMST", "ANTH", "ARAB", "ASTR", "BISC", "CHEM", "CLAS", "COLT", "CORE", "CSLC",
               "EALC", "EASC", "ECON", "ENGL", "ENST", "FREN", "GEOG", "GEOL", "GERM", "SWMS", "GR", "HEBR",
               "HIST", "HBIO", "INDS", "IR", "IRAN", "ITAL", "JS", "LAT", "LING", "MATH", "MDA", "MDES", "MPW",
               "NEUR", "NSCI", "OS", "PHED", "PHIL", "PHYS", "POIR", "PORT", "POSC", "PSYC", "REL", "RNR",
               "SLL", "SOCI", "SPAN", "SSCI", "SSEM", "USC", "VISS", "WRIT", "ACCT", "ARCH", "ACAD", "ACCT",
               "BAEP", "BUAD", "BUCO", "DSO", "FBE", "GSBA", "MKT", "MOR", "HRM", "CMPP", "CNTV", "CTAN",
               "CTCS", "CTIN", "CTPR", "CTWR", "IML", "ASCJ", "CMGT", "COMM", "DSM", "JOUR", "PR", "PUBD",
               "DANC", "DENT", "CBY", "DHIS", "THTR", "EDCO", "EDHP", "EDUC", "AME", "ASTE", "BME", "CHE", "CE",
               "CSCI", "EE", "ENE", "ENGR", "ISE", "INF", "ITP", "MASC", "PTE", "SAE", "ART", "CRIT", "DES",
               "FA", "FACE", "FACS", "FADN", "FADW", "FAIN", "FAPH", "FAPT", "FAPR", "FASC", "WCT", "GCT",
               "SCIN", "SCIS", "ARLT", "SI", "ARTS", "HINQ", "SANA", "LIFE", "PSC", "QREA", "GPG", "GPH",
               "GESM", "GERO", "LAW", "ACMD", "ANST", "BIOC", "CBG", "DSR", "HP", "INTD", "MEDB", "MEDS",
               "MICB", "MPHY", "MSS", "NIIN", "PATH", "PHBI", "PM", "PCPA", "SCRM", "ARTL", "MTEC", "MSCR",
               "MTAL", "MUCM", "MUCO", "MUCD", "MUEN", "MUHL", "MUIN", "MUJZ", "MPEM", "MPGU", "MPKS", "MPPM",
               "MPST", "MPVA", "MPWP", "MUSC", "SCOR", "OT", "HCDA", "PHRD", "PMEP", "RXRS", "BKN", "PT",
               "AEST", "HMGT", "MS", "NAUT", "NSC", "PPD", "PPDE", "PLUS", "RED"]

# Note: 20243 = Fall 2024
semester = 20243

@dataclass
class Section:
    id_and_d_class_code: str
    type: str
    day: str
    time: str
    loc: str
    prof_name: str

@dataclass
class Course:
    name: str
    title: str
    description: str
    units: str
    sections: List[Section]

def getDepartmentJsonString(department_id: str, semester_id: str) -> str:
    url = f"https://web-app.usc.edu/web/soc/api/classes/{department_id}/{semester_id}"
    response = requests.get(url)
    return response.text

def createClassMapping(json_response: str) -> List[Course]:
    try:
        data = json.loads(json_response)
    except json.JSONDecodeError as e:
        print(f"Error parsing JSON: {e}")
        return []

    courses = []
    
    # for all courses from the 'OfferedCourses' section
    for course in data.get("OfferedCourses", {}).get("course", []):
        if not isinstance(course, dict):
            print(f"Warning: Invalid course data: {course}")
            continue
        course_data = course.get("CourseData", {})
        
        # Course Name
        name = f"{course_data.get('prefix', '')}-{course_data.get('number', '')}"
        # Get course title (ie, "Explorations in Computing")
        title = course.get("title", '')
        if (title == ''):
            title = "N/A"
        # Get course description
        description = course.get("description", '')
        if (description == ''):
            description = "N/A"
        # Get units for course (split from "4.0, 0" to "4.0")
        units = (course.get("units", '')).split(",", 1)[0]
        if (units == ''):
            units = "N/A"
        
        # To hold all sections for a course
        sections = []
        for section_data in course_data.get("SectionData", []):
            # Error check to ensure that section is a dictionary
            if not isinstance(section_data, dict):
                print(f"Warning: Invalid section data: {section_data}")
                continue

            id_and_d_class_code = f"{section_data.get('id', '')} {section_data.get('dclass_code', '')}"
            if (id_and_d_class_code == ''):
                id_and_d_class_code = "Unknown"

            type = section_data.get('type', '')
            if (type == ''):
                type = "Unknown"
            
            day = section_data.get('day', '')
            if (day == ''):
                day = "Unknown"

            # Create the time string
            start_time = section_data.get('start_time', '')
            end_time = section_data.get('end_time', '')
            time = f"{start_time} - {end_time}"
            if (time == ''):
                time = "Unknown"

            loc = section_data.get('location', '')
            if (loc == ''):
                loc = "Unknown"
            
            prof = section_data.get('instructor', {})
            if isinstance(prof, dict) and 'first_name' in prof and 'last_name' in prof:
                prof_name = f"{prof['last_name']}, {prof['first_name']}"
            else:
                prof_name = "Unknown"
            
            # Create section
            section = Section(
                id_and_d_class_code=id_and_d_class_code,
                type=type,
                day=day,
                time=time,
                loc=loc,
                prof_name=prof_name
            )
            sections.append(section)

        # Append iff sections found
        if sections:
            course = Course(
                name=name, 
                title=title, 
                description=description,
                units=units,
                sections=sections
            )
            courses.append(course)
    
    return courses

def print_courses(courses: List[Course]):
    if not courses:
        print("No courses found.")
        return

    for course in courses:
        print(f" Name: {course.name}")
        print(f" Title: {course.title}")
        print(f" Description: {course.description}")
        print(f" Units: {course.units}")
        print()
        print("-" * 50)
        for section in course.sections:
            print(f" ID: {section.id_and_d_class_code}")
            print(f" Type: {section.type}")
            print(f" Day: {section.day}")
            print(f" Time: {section.time}")
            print(f" Loc: {section.loc}")
            print(f" Prof Name: {section.prof_name}")
            print()
            print("-" * 50)

if __name__ == "__main__":
    try:
        for department in departments:
            json_response = getDepartmentJsonString(department, semester)
            courses = createClassMapping(json_response)
            print_courses(courses)
    except requests.RequestException as e:
        print(f"Error fetching data from API: {e}")
    except Exception as e:
        import traceback
        print(f"Unexpected error: {e}")
        print(f"Error occurred on line {traceback.extract_tb(e.__traceback__)[-1].lineno}")